### PR TITLE
feat(core): support TypeScript 4.8

### PR DESCRIPTION
**Pull request by crisbeto**
Adds support for TypeScript 4.8 and resolves some issues that came up as a result of the update.

Most of the issues came from some changes in TypeScript where the `decorators` and `modifiers` properties were removed from most node types, and were combined into a single `modifiers` array. Since we need to continue supporting TS 4.6 and 4.7 until v15, I ended up creating a new `ngtsc/ts_compatibility` directory to make it easier to reuse the new backwards-compatible code.

### CODE REVIEW

File: integration/animations/package.json [hunk 1]
```diff
@@ -42,6 +42,6 @@
     "protractor": "file:../../node_modules/protractor",
     "ts-node": "8.3.0",
     "tslint": "5.18.0",
-    "typescript": "file:../../node_modules/typescript"
```
Comment (crisbeto) [inline integration/animations/package.json:45]: These integration tests had to be locked down to 4.7 temporarily, because the CLI needs to be updated.

**Comment by crisbeto** [comment]
[Passing TGP (except for an unrelated flake).](http://test/OCL:465285150:BASE:465311113:1659623955951:6ccb68bd)

File: packages/compiler-cli/src/ngtsc/ts_compatibility/src/ts_cross_version_utils.ts [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import ts from 'typescript';
+
+/** Version of TypeScript parsed to a number for easier comparisons. */
+export const PARSED_TS_VERSION = parseFloat(ts.versionMajorMinor);
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/ts_compatibility/src/ts_cross_version_utils.ts:12]: We shouldn't use `parseFloat` here - it would be very brittle. It'll break, for example, if TS reaches version `4.10`. I'd also be concerned about RC tags and other trailing non-numeric components in the version string. `@angular/compiler-cli` has a dependency already on `semver` - let's use that to parse the version properly.

File: packages/compiler-cli/src/ngtsc/ts_compatibility/src/ts_cross_version_utils.ts [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+
+/** Equivalent of `ts.ModifierLike` which is only present in TS 4.8+. */
+type ModifierLike = ts.Modifier|ts.Decorator;
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/ts_compatibility/src/ts_cross_version_utils.ts:15]: Exported functions use this type, so it should be exported as well.

File: packages/compiler-cli/src/ngtsc/ts_compatibility/src/ts_cross_version_utils.ts [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+
+/**
+ * Updates a `ts.TypeParameter` declaration.
+ *
+ * TODO(crisbeto): this is a backwards-compatibility layer for versions of TypeScript less than 4.7.
+ * We should remove it once we have dropped support for the older versions.
+ */
+export function updateTypeParameterDeclaration(
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/ts_compatibility/src/ts_cross_version_utils.ts:23]: If we're going to create these aliases, we should consider doing the switch at definition time. I'm thinking something like:

```typescript
type Ts47UpdateTypeParameterDeclarationFn = (node: ts.TypeParameterDeclaration, modifiers: ts.Modifier[], name: ts.Identifier, constraint?: ts.TypeNode, defaultType?: ts.TypeNode): ts.TypeParameterDeclaration;

function updateTypeParameterDeclaration_ts47(node: ts.TypeParameterDeclaration, name: ts.Identifier, constraint?: ts.TypeNode, defaultType?: ts.TypeNode): ts.TypeParameterDeclaration {
  return (ts.factory.updateTypeParameterDeclaration as unknown as Ts47UpdateTypeParameterDeclarationFn)(
          node, /* modifiers */[], name, constraint, defaultType);
}

export const updateTypeParameterDeclaration: typeof ts.factory.updateTypeParameterDeclaration = PARSED_TS_VERSION >= 4.8 ? ts.factory.updateTypeParameterDeclaration : updateTypeParameterDeclaration_ts47;
```

This way:
* The switch happens at definition time, allowing us to pass through the call directly to TS in the 4.8 case.
* All other code in the compiler will already be updated to use the 4.8 interface.
* We pass through the type of the function from 4.8, giving us the upstream signature & documentation.
* We declare the type of the 4.7 interface manually (`Ts47UpdateTypeParameterDeclarationFn`).
* We have a wrapper function to translate the 4.8 signature to the 4.7 signature.

**Review by alxhub — CHANGES_REQUESTED** [review]

**Comment by crisbeto** [comment]
I've reworked the PR based on the feedback.

**Comment by crisbeto** [comment]
[New passing TGP (with a couple of flakes)](http://test/OCL:465285150:BASE:466033437:1659967666443:5a1d3d11).

**Comment by crisbeto** [comment]
@alxhub the feedback has been addressed.

**Review by JiaLiPassion — APPROVED** [review]
LGTM!

**Review by Ali-A-Koye — APPROVED** [review]

**Review by josephperrott — APPROVED** [review]
LGTM

File: packages/compiler-cli/src/ngtsc/ts_compatibility/src/ts_cross_version_utils.ts [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+    }
+  }
+
+  return [decorators.length ? decorators : undefined, modifiers.length ? modifiers : undefined];
+}
+
+/** Whether the currently-loaded version of TypeScript is after a specified version. */
+function isAfterTsVersion(targetMajor: number, targetMinor: number): boolean {
+  const [major, minor] = ts.versionMajorMinor.split('.');
+  return parseInt(major) >= targetMajor && parseInt(minor) >= targetMinor;
```
Comment (JoostK) [inline packages/compiler-cli/src/ngtsc/ts_compatibility/src/ts_cross_version_utils.ts:290]: Drive by comment: this will be incorrect for next majors like 5.0, as it fails the minor version check. 

Also note that there's existing code in [`packages/compiler-cli/src/version_helpers.ts`](https://github.com/angular/angular/blob/340fa465ceb166f64a99f06596f11deb4b60240b/packages/compiler-cli/src/version_helpers.ts) that may be usable, but that would likely entail some restructuring of Bazel targets which would be out of scope for this PR.

Comment (crisbeto) [reply packages/compiler-cli/src/ngtsc/ts_compatibility/src/ts_cross_version_utils.ts:290]: Good point, although as I mentioned above, this code should go away relatively soon and it's currently a blocker. I'll switch it back to the `parseFloat` solution for now.

File: packages/compiler-cli/src/ngtsc/core/src/compiler.ts [hunk 6]
```diff
@@ -1107,8 +1108,7 @@ export function isAngularCorePackage(program: ts.Program): boolean {
       return false;
     }
     // It must be exported.
-    if (stmt.modifiers === undefined ||
-        !stmt.modifiers.some(mod => mod.kind === ts.SyntaxKind.ExportKeyword)) {
+    if (!getModifiers(stmt)?.some(mod => mod.kind === ts.SyntaxKind.ExportKeyword)) {
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/core/src/compiler.ts:1111]: I think we could make this more readable without the optional chaining call:

```suggestion
    const modifiers = getModifiers(stmt);
    if (modifiers === undefined || !modifiers.some(mod => mod.kind === ts.SyntaxKind.ExportKeyword)) {
```
Otherwise I think this logic is very hard to follow, with the negation operating on both `undefined` and `false`. Not relying on negation of "falsy" but non-boolean values has been a code style point we've enforced across the compiler.

An alternative would be to make `getModifiers` return an empty array always, which would simplify downstream logic.

File: packages/compiler-cli/src/ngtsc/metadata/src/util.ts [hunk 7]
```diff
@@ -150,14 +151,11 @@ export function extractDirectiveTypeCheckMeta(
...
 function isRestricted(node: ts.Node): boolean {
-  if (node.modifiers === undefined) {
-    return false;
-  }
-
-  return node.modifiers.some(
-      modifier => modifier.kind === ts.SyntaxKind.PrivateKeyword ||
-          modifier.kind === ts.SyntaxKind.ProtectedKeyword ||
-          modifier.kind === ts.SyntaxKind.ReadonlyKeyword);
+  return getModifiers(node)?.some(
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/metadata/src/util.ts:154]: Same here - not relying on optional chaining.

File: packages/compiler-cli/src/ngtsc/reflection/src/typescript.ts [hunk 8]
```diff
@@ -20,11 +22,12 @@ export class TypeScriptReflectionHost implements ReflectionHost {
...
 
   getDecoratorsOfDeclaration(declaration: DeclarationNode): Decorator[]|null {
-    if (declaration.decorators === undefined || declaration.decorators.length === 0) {
-      return null;
-    }
-    return declaration.decorators.map(decorator => this._reflectDecorator(decorator))
-        .filter((dec): dec is Decorator => dec !== null);
+    const decorators = getDecorators(declaration);
+
+    return decorators && decorators.length ?
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/reflection/src/typescript.ts:27]: Same here - use explicit `!== null` comparison rather than relying on falsyness.

File: packages/compiler-cli/src/ngtsc/ts_compatibility/src/ts_cross_version_utils.ts (hunk 2 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/ts_compatibility/src/ts_cross_version_utils.ts:12]: I still see `parseFloat` being used here.

File: packages/compiler-cli/src/ngtsc/ts_compatibility/src/ts_cross_version_utils.ts [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+     name: ts.Identifier, constraint: ts.TypeNode|undefined, defaultType: ts.TypeNode|undefined) =>
+        ts.TypeParameterDeclaration;
+
+/**
+ * Updates a `ts.TypeParameter` declaration.
+ *
+ * TODO(crisbeto): this is a backwards-compatibility layer for versions of TypeScript less than 4.7.
+ * We should remove it once we have dropped support for the older versions.
+ */
+export const updateTypeParameterDeclaration: Ts47UpdateTypeParameterDeclarationFn =
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/ts_compatibility/src/ts_cross_version_utils.ts:29]: I'm a bit confused by the types/logic here.

1) Our compatibility layer should be exposing the TS 4.8 API shape, not the <= 4.7 shape. That is, we should be wrapping older TS versions to be compatible with the new signature, not the other way around.
2) The condition `PARSED_TS_VERSION >= 4.7` seems wrong - shouldn't that be `<=` in this context? Currently it lumps TS 4.7 and 4.8 together, which just seems wrong.

**Review by alxhub — COMMENTED** [review]
As we're doing this, I think we should have an integration test which installs TS 4.7 explicitly and verifies that our compiler still works correctly. We have that coverage now because the CLI is still using 4.7 so most of the tests are pinned, but we should have one that explicitly chooses to use 4.7.

File: packages/compiler-cli/src/ngtsc/ts_compatibility/src/ts_cross_version_utils.ts (hunk 9 above)
Comment (crisbeto) [reply packages/compiler-cli/src/ngtsc/ts_compatibility/src/ts_cross_version_utils.ts:29]: This is a pre-existing utility that I moved out here for consistency: https://github.com/angular/angular/blob/f35f4751b54dcb9e65f2d72aeea87651314159e8/packages/compiler-cli/src/ngtsc/typecheck/src/ts_util.ts#L143

**Comment by crisbeto** [comment]
The latest set of feedback has been addressed. The new code is in the last `fixup` commit.

**Comment by crisbeto** [comment]
I've pushed another commit to update to the 4.8 RC. No other changes were necessary.

**Review by alxhub — APPROVED** [review]
👏 this is looking really good. The compatibility layer is nice and clean, and should be maintainable while we have to deal with this split.

Let's talk on Monday about `ts.getMutableClone` as well.

**Comment by crisbeto** [comment]
Merge assistance: only the inclusive language check is failing on the name of one of the TSLint rule names that we aren't able to change.

**Comment by pkozlowski-opensource** [comment]
This PR was merged into the repository by commit 31429eaccc973672efb4ed98a628cf9842e24b02.
